compiler/rustc_lint/src/if_let_rescope.rs

```diff
@@ -0,0 +1,304 @@
+use std::ops::ControlFlow;
+
+use hir::intravisit::Visitor;
+use rustc_ast::Recovered;
+use rustc_hir as hir;
+use rustc_macros::{LintDiagnostic, Subdiagnostic};
+use rustc_session::lint::FutureIncompatibilityReason;
+use rustc_session::{declare_lint, declare_lint_pass};
+use rustc_span::edition::Edition;
+use rustc_span::Span;
+
+use crate::{LateContext, LateLintPass};
+
+declare_lint! {
+    /// The `if_let_rescope` lint detects cases where a temporary value with
+    /// significant drop is generated on the right hand side of `if let`
+    /// and suggests a rewrite into `match` when possible.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,edition2024
```

Comment:
Problem [REV7]: as correctly pointed out by @ehuss, this is a *migration* lint, so this should not be run only after Edition 2024, as it needs to run on previous editions as well.

Reply:
Suggestion [REV7]: we should also rename PR title and reword description:

- PR title: Rescope temporary lifetime in if-let in Edition 2024 and provide a migration lint
- PR description: mention that we also add a migration lint in this PR, not just the actual semantic changes.

Reply:
I have a question. If we are going to drop `edition2024`, can we drop the `#![feature(if_let_rescope)]` in the lint doc, too? Technically that feature gate is irrelevant unless Edition 2024 is enabled.

Reply:
I would keep the feature gate in the example for now, and only drop the feature gate once this is going to be stabilized (or enabled w/ the edition).